// @6-mils/InboundCxmlMessage/index.js
const debug = require('debug')('6-mils:InboundCxmlMessage')
const { DateTime } = require('luxon')

/**
 * A collection of private property values for each instance of this class.
 * @type {WeakMap}
 */
const _private = new WeakMap()

/**
 * This base class is intended to provide the basic functionality for parsing
 * messages received from a supplier.
 */
class InboundCxmlMessage {
  constructor () {
    /**
     * The properties of the incoming message are set to reasonable defaults.
     * `parse()` will override these values with the ones specified in the cXML
     * message.
     */
    _private.set(this, {
      src: '',
      payloadId: require('@6-mils/CxmlPayloadId')(),
      timestamp: DateTime.local().toString(),
      version: require('@6-mils/CxmlVersion'),
      statusCode: '200',
      statusText: 'success'
    })
  }

  get payloadId () {
    return _private.get(this).payloadId
  }

  get timestamp () {
    return _private.get(this).timestamp
  }

  get version () {
    return _private.get(this).version
  }

  get statusCode () {
    return _private.get(this).statusCode
  }

  get statusText () {
    return _private.get(this).statusText
  }

  /**
   * @param  {String}   src   The cXML message to parse.
   *
   * @return {Promise}  Fulfilled once the parsing is complete.
   */
  parse (src) {
    src = src || ''
    debug('Parsing "%s"', src)

    const self = this
    const props = _private.get(self)

    return new Promise(function (resolve, reject) {
      _private.set(self, props)
      resolve()
    })
  }
}

module.exports = InboundCxmlMessage
